components/tabs/src/storage.rs

```diff
@@ -7,42 +7,122 @@ const URI_LENGTH_MAX: usize = 65536;
 // https://searchfox.org/mozilla-central/rev/ea63a0888d406fae720cf24f4727d87569a8cab5/services/sync/modules/engines/tabs.js#8
 const TAB_ENTRIES_LIMIT: usize = 5;
 
+use crate::error::*;
 use crate::DeviceType;
+use rusqlite::{Connection, OpenFlags};
+use serde_derive::{Deserialize, Serialize};
+use sql_support::open_database::{self, open_database_with_flags};
+use sql_support::ConnExt;
 use std::cell::RefCell;
+use std::path::{Path, PathBuf};
 
-#[derive(Clone, Debug, PartialEq)]
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
 pub struct RemoteTab {
     pub title: String,
     pub url_history: Vec<String>,
     pub icon: Option<String>,
     pub last_used: i64, // In ms.
 }
 
-#[derive(Clone, Debug)]
+#[derive(Clone, Debug, Serialize, Deserialize)]
 pub struct ClientRemoteTabs {
     pub client_id: String, // Corresponds to the `clients` collection ID of the client.
     pub client_name: String,
+    #[serde(
+        default = "devicetype_default_deser",
+        skip_serializing_if = "devicetype_is_unknown"
+    )]
     pub device_type: DeviceType,
     pub remote_tabs: Vec<RemoteTab>,
 }
 
-pub struct TabsStorage {
-    local_tabs: RefCell<Option<Vec<RemoteTab>>>,
-    remote_tabs: RefCell<Option<Vec<ClientRemoteTabs>>>,
+fn devicetype_default_deser() -> DeviceType {
+    // replace with `DeviceType::default_deser` once #4861 lands.
+    DeviceType::Unknown
 }
 
-impl Default for TabsStorage {
-    fn default() -> Self {
-        Self::new()
-    }
+// Unlike most other uses-cases, here we do allow serializing ::Unknown, but skip it.
+fn devicetype_is_unknown(val: &DeviceType) -> bool {
+    matches!(val, DeviceType::Unknown)
+}
+
+// Tabs has unique requirements for storage:
+// * The "local_tabs" exist only so we can sync them out. There's no facility to
+//   query "local tabs", so there's no need to store these persistently - ie, they
+//   are write-only.
+// * The "remote_tabs" exist purely for incoming items via sync - there's no facility
+//   to set them locally - they are read-only.
+// Note that this means a database is only actually needed after Sync fetches remote tabs,
+// and because sync users are in the minority, the use of a database here is purely
+// optional and created on demand. The implication here is that asking for the "remote tabs"
+// when no database exists is considered a normal situation and just implies no remote tabs exist.
```

Comment:
Rewording to ensure I understand, `get_remote_tabs` will retrieve the remote tabs as of the last sync (if any exists) but the consumer should call `sync` if they want to get the absolute latest remote tabs, right?

Reply:
Yep - which is very similar to how things work now, except that before the first sync we previously returned no tabs whereas now we return what was previously persisted.